libraries/service/src/main/java/org/holochain/androidserviceruntime/service/HolochainService.kt

```diff
@@ -249,48 +350,176 @@ class HolochainService : Service() {
 
         // We cannot call Binder.getCallingUid() within the onBind callback,
         // so instead we check the authorization within each IPC call
-        private fun expectAuthorized() {
+        private fun loadClientPackageName() {
+            if (this.clientPackageName != null) {
+                return
+            }
+
             val clientUid = Binder.getCallingUid()
-            val clientPackageName =
+            this.clientPackageName =
                 packageManager.getNameForUid(clientUid)
-                    ?: throw UnauthorizedException(
+                    ?: throw AppBinderUnauthorizedException(
                         "Unable to get name of package binding to AdminBinder uid=$clientUid",
                     )
+        }
 
-            Log.d(
-                logTag,
-                "AppBinder expectAuthorized clientPackageName=$clientPackageName installedAppId=$installedAppId",
-            )
-            var isAuthorized = runtime!!.isAppClientAuthorized(clientPackageName, installedAppId)
+        private fun loadAuthorized() {
+            if (this.authorized) {
+                return
+            }
+            this.authorized = runtime!!.isAppClientAuthorized(this.clientPackageName!!, this.installedAppId)
+        }
 
-            if (!isAuthorized) {
-                Log.d(
-                    logTag,
-                    "authorizeAppClient clientPackageName=$clientPackageName installedAppId=$installedAppId",
-                )
-                // TODO notify user to request authorization
+        private fun isAuthorized(): Boolean {
+            this.loadClientPackageName()
+            this.loadAuthorized()
+            return this.authorized
+        }
 
-                throw UnauthorizedException(
-                    "Package is not authorized to access the AppBinder clientPackageName=$clientPackageName installedAppId=${this.installedAppId}",
-                )
+        private fun requestUserAuthorization(callback: IHolochainServiceCallback) {
+            if (this.isAuthorized()) {
+                return
             }
+
+            // Show notification asking user to authorize this app client
+            HolochainService@ showAppAuthorizationNotification(this.clientPackageName!!, this.installedAppId)
+
+            val exception =
+                AppBinderUnauthorizedException(
+                    "Package is not authorized to access the AppBinder",
+                )
+            callback.appBinderUnauthorizedException(AppBinderUnauthorizedExceptionParcel(exception))
         }
     }
 
-    internal fun authorizeAppClient(
+    /**
```

Comment:
Changes in this file below this line are creating the "request authorization" notification.